scripts/flash_algo.py

```diff
@@ -0,0 +1,350 @@
+        for symbol in self.EXTRA_SYMBOLS:
+            if symbol not in elf_simple.symbols:
+                symbols[symbol] = 0xFFFFFFFF
+            else:
+                symbols[symbol] = elf_simple.symbols[symbol].value
+        self.symbols = symbols
+
+        # Find required sections
+        for section in elf_simple.elf.iter_sections():
+            if bytes2str(section.name) == "PrgCode":
+                if section["sh_type"] == "SHT_PROGBITS":
+                    if ro_section is None:
+                        ro_section = section
+                    else:
+                        raise Exception("Extra ro section")
+                else:
+                    raise Exception("Unexpected section type in PrgCode")
+            if bytes2str(section.name) == "PrgData":
+                if section["sh_type"] == "SHT_PROGBITS":
+                    if rw_section is None:
+                        rw_section = section
+                    else:
+                        raise Exception("Extra rw section")
+                elif section["sh_type"] == "SHT_NOBITS":
+                    if zi_section is None:
+                        zi_section = section
+                    else:
+                        raise Exception("Extra zi section")
+                else:
+                    raise Exception("Unexpected section type in PrgData")
+
+        # Make sure all required sections are present
+        if ro_section is None:
+            raise Exception("Missing ro section")
+        if rw_section is None:
+            raise Exception("Missing rw section")
+        if zi_section is None:
+            zi_section = {
+                "sh_addr": rw_section["sh_addr"] + rw_section["sh_size"],
+                "sh_size": 0,
+            }
+            logger.info("Missing zi section")
+
+        # Build the algo
+        self.ro_start = ro_section["sh_addr"]
+        self.ro_size = ro_section["sh_size"]
+        self.rw_start = rw_section["sh_addr"]
+        self.rw_size = rw_section["sh_size"]
+        self.zi_start = zi_section["sh_addr"]
+        self.zi_size = zi_section["sh_size"]
+
+        # Check section ordering
+        if self.ro_start != 0:
+            raise Exception("RO section does not start at address 0")
+        if self.ro_start + self.ro_size != self.rw_start:
+            raise Exception("RW section does not follow RO section")
+        if self.rw_start + self.rw_size != self.zi_start:
+            raise Exception("ZI section does not follow RW section")
+
+        # Attach data to the flash algo
+        algo_size = self.ro_size + self.rw_size + self.zi_size
+        algo_data = bytearray(algo_size)
+        ro_data = ro_section.data()
+        algo_data[self.ro_start:self.ro_start + self.ro_size] = ro_data
+        rw_data = rw_section.data()
+        algo_data[self.rw_start:self.rw_start + self.rw_size] = rw_data
+        # Note - ZI is already zeroed
+        self.algo_data = algo_data
```

Comment:
Now this is the end of another really giant function, that is indistinguishable from the constructor, because it's only called from the constructor.